compiler/luci/pass/src/FusePreActivationBatchNormPass.cpp

```diff
@@ -0,0 +1,701 @@
+    return nullptr;
+
+  auto opcode = node->opcode();
+  if (opcode == luci::CircleOpcode::CONV_2D)
+  {
+    auto conv = dynamic_cast<luci::CircleConv2D *>(node);
+    auto filter = dynamic_cast<luci::CircleConst *>(conv->filter());
+
+    if (filter == nullptr)
+      return nullptr;
+
+    if (filter->rank() != 4)
+      return nullptr;
+
+    if (filter->dim(0).value() != channel_size)
+      return nullptr;
+
+    if (loco::succs(filter).size() != 1)
+      return nullptr;
+
+    return conv;
+  }
+  else if (opcode == luci::CircleOpcode::MAX_POOL_2D || opcode == luci::CircleOpcode::PAD ||
+           opcode == luci::CircleOpcode::ADD)
+  {
+    auto preds = loco::preds(node);
+    for (auto pred : preds)
+    {
+      auto pred_conv = get_backward_conv2d(static_cast<luci::CircleNode *>(pred), channel_size);
+      if (pred_conv != nullptr)
+        return pred_conv;
+    }
+    return nullptr;
+  }
+  else
+  {
+    return nullptr;
+  }
+}
+
+bool update_conv_bias_with_beta(luci::CircleConv2D *conv, luci::CircleConst *beta, bool add_beta)
+{
+  assert(beta->rank() == 1);
+  auto size = beta->dim(0).value();
+  auto bias = dynamic_cast<luci::CircleConst *>(conv->bias());
+
+  if (bias == nullptr)
+  {
+    bias = conv->graph()->nodes()->create<luci::CircleConst>();
+    bias->dtype(loco::DataType::FLOAT32);
+    bias->rank(1);
+    bias->dim(0).set(size);
+    bias->size<loco::DataType::FLOAT32>(size);
+    conv->bias(bias);
+  }
+  else
+  {
+    if (bias->rank() != 1)
+      return false;
+
+    if (loco::succs(bias).size() != 1)
+      return false;
+
+    if (size != bias->dim(0).value())
+      return false;
+  }
+
+  for (uint32_t i = 0; i < size; i++)
+  {
+    if (add_beta)
+      bias->at<loco::DataType::FLOAT32>(i) += beta->at<loco::DataType::FLOAT32>(i);
+    else
+      bias->at<loco::DataType::FLOAT32>(i) -= beta->at<loco::DataType::FLOAT32>(i);
+  }
+  return true;
+}
+
+void update_conv_weights_with_gamma(luci::CircleConv2D *conv, luci::CircleConst *gamma)
+{
+  auto filter = dynamic_cast<luci::CircleConst *>(conv->filter());
+  assert(filter != nullptr);
+
+  uint32_t filter_out_dim = filter->dim(0).value();
+  uint32_t filter_height_dim = filter->dim(1).value();
+  uint32_t filter_width_dim = filter->dim(2).value();
+  uint32_t filter_in_dim = filter->dim(3).value();
+  for (uint32_t o = 0; o < filter_out_dim; o++)
+  {
+    for (uint32_t h = 0; h < filter_height_dim; h++)
+    {
+      for (uint32_t w = 0; w < filter_width_dim; w++)
+      {
+        for (uint32_t i = 0; i < filter_in_dim; i++)
+        {
+          uint32_t offset = o * filter_height_dim * filter_width_dim * filter_in_dim +
+                            h * filter_width_dim * filter_in_dim + w * filter_in_dim + i;
+          filter->at<loco::DataType::FLOAT32>(offset) *= gamma->at<loco::DataType::FLOAT32>(i);
+        }
+      }
+    }
+  }
+}
+
+luci::CircleSub *insert_sub(luci::CircleNode *pred, luci::CircleConst *beta)
```

Comment:
`const luci::CircleNode *pred` and `const luci::CircleConst *beta` ?

Reply:
Changing them to const generates compilation error (because of no interface implementation).